### PR TITLE
Post Put Delete And Image and File Content Responses

### PR DESCRIPTION
**Pull request by tbaik**

### CODE REVIEW

File: src/com/tony/server/Main.java [hunk 1]
```diff
@@ -1,8 +1,6 @@
 package com.tony.server;
 
-import com.tony.server.response.GetResponse;
-import com.tony.server.response.OptionsResponse;
-import com.tony.server.response.PutPostResponse;
+import com.tony.server.response.*;
```
Comment (tk-kowal) [inline src/com/tony/server/Main.java]: Don't use wildcard imports. Be explicit about what you're importing even if that means three lines of imports vs one.

File: src/com/tony/server/Main.java [hunk 2]
```diff
@@ -38,16 +36,22 @@ public static void main(String[] args) {
...
+        router.addRoute(new Request("GET", "/"), new FileContentResponse(directoryPath + "/"));
+        router.addRoute(new Request("GET", "/form"), new FileContentResponse(directoryPath + "/form"));
+        router.addRoute(new Request("POST", "/form"), new PutPostResponse(directoryPath + "/form"));
+        router.addRoute(new Request("PUT", "/form"), new PutPostResponse(directoryPath + "/form"));
+        router.addRoute(new Request("DELETE", "/form"), new DeleteResponse(directoryPath + "/form"));
+        router.addRoute(new Request("GET", "/method_options"), new FileContentResponse(directoryPath + "/method_options"));
+        router.addRoute(new Request("HEAD", "/method_options"), new HeadResponse());
+        router.addRoute(new Request("POST", "/method_options"), new PutPostResponse(directoryPath + "/method_options"));
+        router.addRoute(new Request("PUT", "/method_options"), new PutPostResponse(directoryPath + "/method_options"));
         router.addRoute(new Request("OPTIONS", "/method_options"), new OptionsResponse(router, "/method_options"));
```
Comment (tk-kowal) [inline src/com/tony/server/Main.java]: No need to pass the entire router to OptionsResponse here, it only needs the allowed methods so pass those in instead

File: src/com/tony/server/Main.java [hunk 3]
```diff
@@ -38,16 +36,22 @@ public static void main(String[] args) {
...
+        router.addRoute(new Request("GET", "/file1"), new FileContentResponse(directoryPath + "file1"));
+        router.addRoute(new Request("GET", "/image.jpeg"), new ImageContentResponse(directoryPath + "image.jpeg"));
+        router.addRoute(new Request("GET", "/image.png"), new ImageContentResponse(directoryPath + "image.png"));
+        router.addRoute(new Request("GET", "/image.gif"), new ImageContentResponse(directoryPath + "image.gif"));
```
Comment (tk-kowal) [inline src/com/tony/server/Main.java]: See if you can refactor this to be a bit cleaner. You're repeating yourself here quite a bit (a route needs the path specified for the request as well as the response).

File: src/com/tony/server/response/DeleteResponse.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    public DeleteResponse(String filePath) {
+        this.filePath = filePath;
+        setStatusLine("HTTP/1.1 200 OK\n");
+        setHeaders(new HashMap<>());
+        setBody("");
+    }
+
+    public byte[] respond(){
+        new File(filePath).delete();
```
Comment (tk-kowal) [inline src/com/tony/server/response/DeleteResponse.java]: what happens if the delete fails?

File: src/com/tony/server/response/FileContentResponse.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import java.nio.file.Files;
+import java.util.HashMap;
+
+public class FileContentResponse extends Response{
+    private final String filePath;
+
+    public byte[] respond(){
+        try {
+            if(new File(filePath).isFile()){
+                setBody(new String(Files.readAllBytes(new File(filePath).toPath())));
```
Comment (tk-kowal) [inline src/com/tony/server/response/FileContentResponse.java]: You're new-ing up a File twice here.

File: src/com/tony/server/response/ImageContentResponse.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+public class ImageContentResponse extends Response{
+
+    private final String filePath;
+
+    public byte[] respond(){
+        try {
+            byte[] imageBytes = Files.readAllBytes(Paths.get(filePath));
+            return ResponseBuilder.buildImageResponse(getStatusLine(), getHeaders(), imageBytes);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (tk-kowal) [inline src/com/tony/server/response/ImageContentResponse.java]: I haven't actually tested this. Will I get a 500 response if something goes wrong here? Will I get a 500 if there's any internal server error?

File: src/com/tony/server/response/PutPostResponse.java [hunk 7]
```diff
@@ -1,11 +1,45 @@
...
+        putPostContent();
+        return ResponseBuilder.buildResponse(getStatusLine(),
+                getHeaders(), getBody()).getBytes();
+    }
+
+    public void putPostContent() {
+        try {
+            PrintWriter writer = new PrintWriter(filePath, "UTF-8");
+            writer.print(requestBody);
+            writer.close();
```
Comment (tk-kowal) [inline src/com/tony/server/response/PutPostResponse.java]: Can you think of a situation where this 'writer.close()' might not be executed? Search for 'try-with-resource' and 'finally' blocks in Java.

File: tests/com/tony/server/MainTest.java [hunk 8]
```diff
@@ -1,43 +1,45 @@
...
 
 import java.util.ArrayList;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
 
 public class MainTest {
     @Test
     public void testCreateURIListHasAllURIFromDirectory() throws Exception {
         String directory = System.getProperty("user.dir") + "/public/";
```
Comment (tk-kowal) [inline tests/com/tony/server/MainTest.java]: This test is pretty fragile in that it depends on certain files actually existing in the public folder. What are some other options here?

File: tests/com/tony/server/response/ImageContentResponseTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package com.tony.server.response;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class ImageContentResponseTest {
+    @Test
+    public void testResponseHasCorrectContentTypeJPEGWhenGivenJPEGFile() throws Exception {
```
Comment (tk-kowal) [inline tests/com/tony/server/response/ImageContentResponseTest.java]: I'm curious, does this work if the image type is *.jpg instead of *.jpeg?

Comment (tbaik) [inline tests/com/tony/server/response/ImageContentResponseTest.java]: Yes, it works!

File: src/com/tony/server/response/DeleteResponse.java (hunk 4 above)
Comment (tbaik) [inline src/com/tony/server/response/DeleteResponse.java]: It returns a 404 if there's no route to the file. It's supposed to be a 200 if the file to be deleted is already deleted

**Pull request merged**